uPortal-groups/uPortal-groups-smartldap/src/main/java/org/apereo/portal/groups/smartldap/SmartLdapGroupStore.java

```diff
@@ -235,53 +240,72 @@ public IEntityGroup find(String key) throws GroupsException {
+                rslt.addAll(getParentGroups(s, new HashSet<>()));
+            }
+        }
+        return rslt.iterator();
+    }
 
-            // Ask the individual...
-            EntityIdentifier ei = gm.getUnderlyingEntityIdentifier();
-            IPersonAttributes attr = personAttributeDao.getPerson(ei.getKey());
-            // avoid NPEs and unnecessary IPerson creation
-            if (attr != null && attr.getAttributes() != null && !attr.getAttributes().isEmpty()) {
-                IPerson p = PersonFactory.createPerson();
-                p.setAttributes(attr.getAttributes());
-
-                // Analyze its memberships...
-                Object[] groupKeys = p.getAttributeValues(memberOfAttributeName);
-                // IPerson returns null if no value is defined for this attribute...
-                if (groupKeys != null) {
-
-                    List<String> list = new ArrayList<>();
-                    for (Object o : groupKeys) {
-                        list.add((String) o);
-                    }
+    // gm should already be determined to be reference to person
+    private Object[] getPersonGroupMemberKeys(IGroupMember gm) {
+        Object[] keys = null;
+        EntityIdentifier ei = gm.getUnderlyingEntityIdentifier();
+        IPersonAttributes attr = personAttributeDao.getPerson(ei.getKey());
+        if (attr != null && attr.getAttributes() != null && !attr.getAttributes().isEmpty()) {
+            IPerson p = PersonFactory.createPerson();
+            p.setAttributes(attr.getAttributes());
+            keys = p.getAttributeValues(memberOfAttributeName);
+            log.debug("Groups for person {} is: {}", p.getUserName(), Arrays.toString(keys));
+        }
+        return keys != null ? keys : new Object[] {};
+    }
 
-                    for (String s : list) {
-                        if (groupsTree.getGroups().containsKey(s)) {
-                            rslt.add(groupsTree.getGroups().get(s));
-                        }
-                    }
-                }
+    private Set<IEntityGroup> getParentGroups(String key, Set<IEntityGroup> groups) {
+        // groups is an ongoing collection to avoid recursion
+        log.debug("Getting parents of group: {}", key);
+        IEntityGroup group = groupsTree.getGroups().get(key);
+        if (group == null) {
+            log.warn("SmartLdap group not found for key: {}", key);
+            return groups;
+        }
+        List<String> parentKeys = groupsTree.getParents().get(key);
+        parentKeys = parentKeys != null ? parentKeys : Collections.emptyList();
+        log.debug("Parent keys for {}: {}", key, String.join(",", parentKeys));
+        for (String parentKey : parentKeys) {
+            IEntityGroup parent = groupsTree.getGroups().get(parentKey);
+            if (parent == null) {
+                log.warn("Group tree inconsistent -- missing parent: {}", parentKey);
+            } else if (groups.contains(parent)) {
+                log.warn(
```

Comment:
This log could be an info level because several childs will have the same parent (in a group tree structure).
Even I don't know if there are case which will need to be logged, this part can be removed and next else should test if `!groups.contains(parent)`

Reply:
@bjagg : Is there a possibility to review this log ?
I would like to keep the info level on this class without a big number of this warn log (this warn is normal in my case as we make group tree decomposition from child and that we come back to parent and that all childs have a common parent, so there isn't need to warn about existing parent `groups.contains(parent)`)